### DISCUSSION TITLE
Location filter and column improvement

### DISCUSSION THREAD

**Discussion by celogeek** [jesec/flood#251] [Ideas]
The path system seems strange.

For single files with Torrent Location set to "/data", I've got "/data/mymovie.mkv" and the location column shows "/data"
For multi files, I've got "/data/mymovie/file1.nfo" and "/data/mymovie/file1.mkv" and the location column shows "/data/mymovie" instead of "/data".

When I want to move the torrent, this location is shown, and we would think we need to keep the /mymovie part on the new location, but no.
if I set change "/data/mymovie/" (which is what set torrent location show when we what to move it), and I change the /data to /newdata, then I would think it will move the data to "/newdata/mymovie" but in fact the files end up in "/newdata/mymovie/mymovie"

Idea of improvement:
 - the single file stay the same
 - the multi files show display the base path instead of the torrent path
 - the set torrent location for multi files should purpose the base path instead of the multifiles
 - remove the use "base path" option (not sure what it does, this should be transparent for the user)
 - purpose filters based on the path (and may be clear the most common part of them
   - for example if I have "/data/Movies", "/data/Docs", "/data/Apps", show a filter with "Movies, Docs, Apps".

Alternatively,
 - purpose a columns with the base path
 - preference to use base path or location
 - filter on the base path

**Comment by jesec** [comment]
This is the expected behavior.

Suppose the destination is `/Downloads`. 

When you add/move a multi-file torrent that has the following structure:

```
Big buck bunny (torrent name)
      - BBB.mkv
      - BBB.ass
```

when `Base Path` is not toggled (default), you got:

```
/Downloads
      - Big buck bunny
            - BBB.mkv
            - BBB.ass
```

when `Base Path` is toggled, you got:

```
/Downloads
      - BBB.mkv
      - BBB.ass
```

`Base Path` is not a concern to single-file torrents.

The final results of add/move are always defined and reproducible, given the same options. 

Meanwhile, `location` column may or may not be the base path, and users should not make assumptions. 

IMO no change should be made.

**Comment by celogeek** [comment]
So the base bath of a multi files is a subdirectory.

Is it possible to have a column with the Parent Location? The value would be the same for single file and multi files?

It is possible to allow tree filter on Parent Location ?

So on the left (like tracks for instance), something like:

- data
  - downloads
    - New
    - Movies
    - Series
    - Docs

I can click on it, and I can see only the files saved here.

**Reply by sabersalv** [reply]
@celogeek I like your idea. Transmission Web Control has [this feature](https://github.com/ronggang/transmission-web-control), In this way, can use folders to organize torrents, along with tags for extra filtering

**Reply by celogeek** [reply]
Thanks !

And it is important to show only the base_path (or parent ?) for multi files.

So a multifiles will have /data/downloads/movies/mymovies/...
And will show in the filter /data/downloads/movies

Otherwise the tree will be very heavy.

**Reply by jesec** [reply]
Unfortunately this can not be done. Torrent clients usually don't tell you whether the directory is the base path or not in the torrent list API response, as there are a dozen of special cases: single-file vs. multi-file, options used during add/move, directory rename, etc.

To get the actual path to contents, usually a second API call has to be issued, which is time-consuming and resource-intensive. That's why things like trackers/peers/contents are in a separate "Details" modal. 

Generally, you can use the `directory` column to organize your torrents just fine.

**Reply by celogeek** [reply]
You mean "location" columns?

I take a look at what the xmlrpc returns:

For a single file:
d.base_path => /data/download/Manga/x.mkv
d.directory_base => /data/download/Manga

For a multi files:
d.base_path => d.directory_base

we can have:
if d.base_path == d.directory_base return dirname(d.base_path)
else return d.directory_base

Anyway in both case a dirname(d.base_path) will return the parent directory anyway.
